libminifi/include/io/Stream.h

```diff
@@ -0,0 +1,52 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace io {
+
+/**
+ * All streams serialize/deserialize in big-endian
+ */
+class Stream {
+ public:
+  virtual void close() {}
+
+  virtual void seek(uint64_t offset) {
+    throw std::runtime_error("Seek is not supported");
+  }
+
+  virtual int initialize() {
+    return 1;
+  }
+
+  virtual const uint8_t* getBuffer() const {
+    throw std::runtime_error("Not a buffered stream");
+  }
+  virtual ~Stream() = default;
+};
```

Comment:
I'm not entirely sure where we should put them, for example `FileOutputCallback::process` takes a `BaseStream` and then calls `getBuffer` (which only makes sense for a `BufferStream`), if we would like to change the method, that it can only take a `BufferStream`, then we are up for a (not necessary infeasible) ride, I moved these methods back and forth between this Stream class and the BaseStream, `ZlibCompressStream` is only an `OutputStream` but it does implement a non-trivial `close` method, that's why they settled in `Stream`, also in the future `content_repo->write(claim)` should only return an `OutputStream`, not a `BaseStream`, and as it stands we do call `seek` on the result

I think we should investigate further with your proposed "granularization" in a separate PR